sys/kern/kern_proc.c

```diff
@@ -2803,44 +2835,67 @@ kern_proc_vmmap_out(struct proc *p, struct sbuf *sb, ssize_t maxlen, int flags)
 			kve->kve_type = KVME_TYPE_NONE;
 			kve->kve_ref_count = 0;
 			kve->kve_shadow_count = 0;
+
+			/* Anonymous memory is presumed able to have tags */
+			kve->kve_protection |= KVME_PROT_LOADTAGS
+					       | KVME_PROT_STORETAGS;
+
+			kve->kve_max_protection |= KVME_PROT_LOADTAGS
+						   | KVME_PROT_STORETAGS;
 		}
 
-		strlcpy(kve->kve_path, fullpath, sizeof(kve->kve_path));
+		if ((kve->kve_protection & KVME_PROT_READ) == 0)
+			kve->kve_protection &= ~KVME_PROT_LOADTAGS;
+		if ((kve->kve_protection & KVME_PROT_WRITE) == 0)
+			kve->kve_protection &= ~KVME_PROT_STORETAGS;
+		if ((kve->kve_max_protection & KVME_PROT_READ) == 0)
+			kve->kve_max_protection &= ~KVME_PROT_LOADTAGS;
+		if ((kve->kve_max_protection & KVME_PROT_WRITE) == 0)
+			kve->kve_max_protection &= ~KVME_PROT_STORETAGS;
+
+		pathlen = strlcpy(kve->kve_path, fullpath, sizeof(kve->kve_path));
 		if (freepath != NULL)
 			free(freepath, M_TEMP);
 
 		/* Pack record size down */
-		if ((flags & KERN_VMMAP_PACK_KINFO) != 0)
+		if ((flags & KERN_VMMAP_PACK_KINFO) != 0) {
+			if (pathlen > sizeof(kve->kve_path))
+				pathlen = sizeof(kve->kve_path);
+
 			kve->kve_structsize =
 			    offsetof(struct kinfo_vmentry, kve_path) +
-			    strlen(kve->kve_path) + 1;
-		else
+			    pathlen + 1;
+		} else
 			kve->kve_structsize = sizeof(*kve);
 		kve->kve_structsize = roundup(kve->kve_structsize,
 		    sizeof(uint64_t));
 
 		/* Halt filling and truncate rather than exceeding maxlen */
 		if (maxlen != -1 && maxlen < kve->kve_structsize) {
 			error = 0;
-			vm_map_lock_read(map);
-			break;
+			goto out_nolock;
 		} else if (maxlen != -1)
 			maxlen -= kve->kve_structsize;
 
 		if (sbuf_bcat(sb, kve, kve->kve_structsize) != 0)
 			error = ENOMEM;
-		vm_map_lock_read(map);
 		if (error != 0)
-			break;
+			goto out_nolock;
+		vm_map_lock_read(map);
 		if (last_timestamp != map->timestamp) {
 			vm_map_lookup_entry(map, addr - 1, &tmp_entry);
 			entry = tmp_entry;
 		}
 	}
 	vm_map_unlock_read(map);
+out_nolock:
 	vmspace_free(vm);
 	PRELE(p);
 	free(kve, M_TEMP);
+
+	if (i == 0)
+		error = ENOENT;
```

Comment:
Asking for nothing and getting nothing back isn't necessarily an error?

Reply:
That's fair!  Not sure why I wrote that.